packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/records/RecordTypeConverter.kt

```diff
@@ -3,26 +3,28 @@ package expo.modules.kotlin.records
 import com.facebook.react.bridge.Dynamic
 import expo.modules.kotlin.allocators.ObjectConstructor
 import expo.modules.kotlin.allocators.ObjectConstructorFactory
-import expo.modules.kotlin.types.KClassTypeWrapper
 import expo.modules.kotlin.types.TypeConverter
-import expo.modules.kotlin.types.TypeConverterHelper
+import expo.modules.kotlin.types.TypeConverterProvider
+import kotlin.reflect.KClass
+import kotlin.reflect.KType
 import kotlin.reflect.full.findAnnotation
-import kotlin.reflect.full.isSuperclassOf
 import kotlin.reflect.full.memberProperties
 import kotlin.reflect.jvm.javaField
 
-class RecordTypeConverter : TypeConverter {
+// TODO(@lukmccall): create all converters during initialization
+class RecordTypeConverter<T : Record>(
+  private val converterProvider: TypeConverterProvider,
+  val type: KType,
+) : TypeConverter<T>(type.isMarkedNullable) {
   private val objectConstructorFactory = ObjectConstructorFactory()
 
-  override fun canHandleConversion(toType: KClassTypeWrapper): Boolean =
-    Record::class.isSuperclassOf(toType.classifier)
+  override fun notOptionalConvert(value: Dynamic): T {
```

Comment:
Nitpick 😅
1. "not-optional" sounds less natural than "non-optional" when it's an adjective. Not only here, but everywhere in this PR code
2. IMO inverted word order is more descriptive (the converted value is non-optional, not the conversion itself)

```suggestion
  override fun convertNonOptional(value: Dynamic): T {
```